File: SZE/OODB/kyberszittya/Rec4/Recitation4/Labirintus/unit.cpp
#include "Unit.h"
#include "Padloelem.h"

#include <iostream>

bool Unit::move(MovingDirection movedir)
{
	PadloElem* p = current_padloelem->getNeighborElem(movedir);
	if (p != NULL)
	{
		if (!p->isFoglalt())
		{
			current_padloelem->setOccupyingUnit(NULL);
			setCurrentPadloelem(current_padloelem->getNeighborElem(movedir));
			return true;
		}
		else if (p->getOccupyingUnit()->getUnitType() != UnitType::OBSTACLE)
		{
			p->getOccupyingUnit()->damage(this->unitstat.attack, this->unitstat.damage);
			std::cout << "Attacking monster" << std::endl;
			return true;
		}
	}
	else
	{
		return false;
	}
}

void Unit::damage(unsigned int attack, unsigned int damage)
{
	unsigned int dmg = (unsigned int)(((double)(unitstat.defense) / (double)(unitstat.attack)) * damage);
	if (unitstat.health <= dmg)
	{
		unitstat.health = 0;
		kill();
	}
	else
	{
		unitstat.health -= (unsigned int)(((double)(unitstat.defense) / (double)(unitstat.attack)) * damage);
	}	
	std::cout << unitstat.health << std::endl;
}

void Unit::setCurrentPadloelem(PadloElem* padloelem)
{
	current_padloelem = padloelem;
	current_padloelem->setOccupyingUnit(this);
}

bool Unit::kill()
{
	if (isKilled())
	{
		std::cout << "DEAD" << std::endl;
		if (!dead)
		{
			
			current_padloelem->setOccupyingUnit(NULL);
			current_padloelem = NULL;
			dead = true;
		}
		return true;
	}
	return false;
}

File: Misc/RPG_projekt/src/0110_constructor_visibility_files/Unit.java
class Unit {
    private String name;
    private int health;
    private int damage;
    private int defense;

    public Unit(String name, int health, int damage, int defense) {
        this.name = name;
        this.health = health;
        this.damage = damage;
        this.defense = defense;
    } 

    public boolean isAlive() {
        return this.health > 0;
    }

    public void attack(Unit other) {
        int damage = this.damage > other.defense ? this.damage - other.defense : 0;
        other.health -= damage;
        if (other.health < 0) {
            other.health = 0;
        }
    }
    
    public String getDisplayText() {
        return String.format("%s (HP: %d, DMG: %d, DEF: %d)", this.name, this.health, this.damage, this.defense);
    }
}